### ISSUE TITLE
FDs sent by Libvirt are being reset in Cloud-Hypervisor.

### ISSUE THREAD

**Issue by praveen-pk** [cloud-hypervisor/cloud-hypervisor#5631]
**Describe the bug**
Libvirt sends FDs of the TAP devices created on the host to CH (Cloud-hypervisor) via HTTP request. These FDs are being reset starting with **v33**.

**To Reproduce**
The Libvirt in question is not yet merged upstream. https://github.com/praveen-pk/libvirt/tree/mshv is the branch which can be used to reproduce the issue.
 
```
<domain type='kvm'>
  <name>clh_net</name>
  <title>Clh Net Test</title>
  <memory unit='G'>2</memory>
  <os>
          <type>hvm</type>
          <kernel>hypervisor-fw</kernel>
  </os>
  <devices>
      <disk type='file'>
          <source file='guest_disk.raw'/>
            <driver if='virtio' type='raw' queues='4'/>
            <target dev='vda'/>
    </disk>
        <serial type="pty">
            <target port="0"></target>
        </serial>
  <interface type='ethernet'>
      <mac address='aa:bb:cc:b0:72:68'/>
      <target dev='tap1' managed='no'/>
   </interface>

 </devices>
    <vcpu placement="static">2</vcpu>

</domain>
```

This guest XML causes libvirt to send below net settings to guest via HTTP:
```
"net":[{"tap":null,"ip":"192.168.249.1","mask":"255.255.255.0","mac":"aa:bb:cc:b0:72:68","host_mac":null,"mtu":null,"iommu":false,"num_queues":2,"queue_size":256,"vhost_user":false,"vhost_socket":null,"vhost_mode":"Client","id":"_net2","fds":[25],"rate_limiter_config":null,"pci_segment":0,"offload_tso":true,"offload_ufo":true,"offload_csum":true}]
```
After 
https://github.com/cloud-hypervisor/cloud-hypervisor/commit/fba0b5f9

the FDs in create request are being reset. Because of this, guest is not able to use the passed FD and so fails to connect to specified network.

**Version**

cloud-hypervisor v33.0 (Built from sources)

**VM configuration**

Output of vm.info from the guest with FDs being reset

```
{"config":{"cpus":{"boot_vcpus":2,"max_vcpus":2,"topology":null,"kvm_hyperv":false,"max_phys_bits":46,"affinity":null,"features":{"amx":false}},"memory":{"size":2147483648,"mergeable":false,"hotplug_method":"Acpi","hotplug_size":null,"hotplugged_size":null,"shared":false,"hugepages":false,"hugepage_size":null,"prefault":false,"zones":null,"thp":true},"payload":{"firmware":null,"kernel":"/vms/hypervisor-fw","cmdline":null,"initramfs":null},"disks":[{"path":"/vms/core-2.0.20230321.raw","readonly":false,"direct":false,"iommu":false,"num_queues":1,"queue_size":128,"vhost_user":false,"vhost_socket":null,"rate_limiter_config":null,"id":"_disk0","disable_io_uring":false,"pci_segment":0},{"path":"/vms/cloud-config-dhcp-ens4.raw","readonly":false,"direct":false,"iommu":false,"num_queues":1,"queue_size":128,"vhost_user":false,"vhost_socket":null,"rate_limiter_config":null,"id":"_disk1","disable_io_uring":false,"pci_segment":0}],"net":[{"tap":null,"ip":"192.168.249.1","mask":"255.255.255.0","mac":"aa:bb:cc:b0:72:68","host_mac":"76:23:59:e8:c1:27","mtu":null,"iommu":false,"num_queues":2,"queue_size":256,"vhost_user":false,"vhost_socket":null,"vhost_mode":"Client","id":"_net2","fds":null,"rate_limiter_config":null,"pci_segment":0,"offload_tso":true,"offload_ufo":true,"offload_csum":true}],"rng":{"src":"/dev/urandom","iommu":false},"balloon":null,"fs":null,"pmem":null,"serial":{"file":"/dev/pts/8","mode":"Pty","iommu":false},"console":{"file":null,"mode":"Tty","iommu":false},"devices":null,"user_devices":null,"vdpa":null,"vsock":null,"iommu":false,"sgx_epc":null,"numa":null,"watchdog":false,"platform":null,"tpm":null},"state":"Running","memory_actual_size":2147483648,"device_tree":{"_virtio-pci-__console":{"id":"_virtio-pci-__console","resources":[{"PciBar":{"index":0,"base":70364448686080,"size":524288,"type_":"Mmio64","prefetchable":false}}],"parent":null,"children":["__console"],"pci_bdf":"0000:00:01.0"},"_virtio-pci-_disk1":{"id":"_virtio-pci-_disk1","resources":[{"PciBar":{"index":0,"base":3891265536,"size":524288,"type_":"Mmio32","prefetchable":false}}],"parent":null,"children":["_disk1"],"pci_bdf":"0000:00:03.0"},"__serial":{"id":"__serial","resources":[],"parent":null,"children":[],"pci_bdf":null},"_virtio-pci-__rng":{"id":"_virtio-pci-__rng","resources":[{"PciBar":{"index":0,"base":70364447637504,"size":524288,"type_":"Mmio64","prefetchable":false}}],"parent":null,"children":["__rng"],"pci_bdf":"0000:00:05.0"},"_virtio-pci-_net2":{"id":"_virtio-pci-_net2","resources":[{"PciBar":{"index":0,"base":70364448161792,"size":524288,"type_":"Mmio64","prefetchable":false}}],"parent":null,"children":["_net2"],"pci_bdf":"0000:00:04.0"},"_net2":{"id":"_net2","resources":[],"parent":"_virtio-pci-_net2","children":[],"pci_bdf":null},"__console":{"id":"__console","resources":[],"parent":"_virtio-pci-__console","children":[],"pci_bdf":null},"_disk1":{"id":"_disk1","resources":[],"parent":"_virtio-pci-_disk1","children":[],"pci_bdf":null},"_virtio-pci-_disk0":{"id":"_virtio-pci-_disk0","resources":[{"PciBar":{"index":0,"base":3891789824,"size":524288,"type_":"Mmio32","prefetchable":false}}],"parent":null,"children":["_disk0"],"pci_bdf":"0000:00:02.0"},"_disk0":{"id":"_disk0","resources":[],"parent":"_virtio-pci-_disk0","children":[],"pci_bdf":null},"__rng":{"id":"__rng","resources":[],"parent":"_virtio-pci-__rng","children":[],"pci_bdf":null},"__ioapic":{"id":"__ioapic","resources":[],"parent":null,"children":[],"pci_bdf":null}}}
```

**Comment by praveen-pk** [comment]
@alyssais, could you provide some details on why the FDs are being reset in CH? I see you pushed above listed patch. Couldn't find related Bug with context.

**Comment by rbradford** [comment]
libvirt needs to use `SCM_RIGHTS` to send the fd (via `sendmsg`) rather than specifying an open FD.

**Comment by liuw** [comment]
@praveen-pk see this https://github.com/cloud-hypervisor/cloud-hypervisor/security/advisories/GHSA-g6mw-f26h-4jgp

**Comment by praveen-pk** [comment]
Thanks for the context @liuw. 

I researched using SCM_RIGHTS a bit, I see some missing links.

In current implementation, FDs of TAP devices are passed to cloud-hypervisor at the time CH Process is forked from Libvirt. The FD# (FD Index) recorded at the time of CH Process forking and is passed in `Vm.create` request. At this point, the FD is already passed to CH process and closed in Libvirt process. Only an index(FD#) is passed via HTTP `vm.create` request. This FD# is used in `virtio_devices::Net::from_tap_fds` to create a TAP device in Cloud-Hypervisor.

I could use `SCM_RIGHTS` to send FD to Cloud-hypervisor. Libvirt still needs to communicate which FD#s to use while creating TAP devices in Cloud-hypervisor. This is a missing link to go down SCM_RIGHTS path. This would get further complicated if Libvirt were to pass multiple TAP devices to Cloud-Hypervisor.

To summarize. even if FDs were to be passed to Cloud-Hypervisor via `SCM_RIGHTS`, libvirt needs a mechanism to specify which tap FD# should be used while creating a virtio-nic.


Did I miss something here?

**Comment by rbradford** [comment]
You need to do separate `VmCreate` (without network) and then `AddNet`. When you call `AddNet` you can pass FDs using `sendmsg`

**Comment by rbradford** [comment]
This is not a CH issue - the libvirt implementation needs to be adapted to follow the correct behaviour.
